sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java

```diff
@@ -0,0 +1,565 @@
+            );
+        } catch (IOException e) {
+            throw ExceptionsHelper.convertToOpenSearchException(e);
+        }
+    }
+
+    protected static class ConfigUpdatingActionListener<Response> implements ActionListener<Response> {
+        private final String[] cTypes;
+        private final Client client;
+        private final ActionListener<Response> delegate;
+
+        public ConfigUpdatingActionListener(String[] cTypes, Client client, ActionListener<Response> delegate) {
+            this.cTypes = Objects.requireNonNull(cTypes, "cTypes must not be null");
+            this.client = Objects.requireNonNull(client, "client must not be null");
+            this.delegate = Objects.requireNonNull(delegate, "delegate must not be null");
+        }
+
+        @Override
+        public void onResponse(Response response) {
+
+            final ConfigUpdateRequest cur = new ConfigUpdateRequest(cTypes);
+
+            client.execute(ConfigUpdateAction.INSTANCE, cur, new ActionListener<>() {
+                @Override
+                public void onResponse(final ConfigUpdateResponse ur) {
+                    if (ur.hasFailures()) {
+                        delegate.onFailure(ur.failures().get(0));
+                        return;
+                    }
+                    delegate.onResponse(response);
+                }
+
+                @Override
+                public void onFailure(final Exception e) {
+                    delegate.onFailure(e);
+                }
+            });
+
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            delegate.onFailure(e);
+        }
+
+    }
+
+    @Override
+    protected final RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+
+        // consume all parameters first so we can return a correct HTTP status,
+        // not 400
+        consumeParameters(request);
+
+        // check if .opendistro_security index has been initialized
+        if (!ensureIndexExists()) {
+            return channel -> internalErrorResponse(channel, ErrorType.SECURITY_NOT_INITIALIZED.getMessage());
+        }
+
+        // check if request is authorized
+        String authError = restApiPrivilegesEvaluator.checkAccessPermissions(request, getEndpoint());
+
+        final User user = (User) threadPool.getThreadContext().getTransient(ConfigConstants.IDENTITY_USER);
+        final String userName = user == null ? null : user.getUsername().getName();
+        if (authError != null) {
+            log.error("No permission to access REST API: " + authError);
+            // for rest request
+            request.params().clear();
+            return channel -> forbidden(channel, "No permission to access REST API: " + authError);
+        }
+
+        final Object originalUser = threadPool.getThreadContext().getTransient(ConfigConstants.IDENTITY_USER);
```

Comment:
@peternied Any thoughts on how interactions with ThreadContext should work in the identity module?